File: _legacy/HWFlow_legacy/python/parser_docx.py
"""
parser_docx.py — .docx 파일을 중간표현(IR)으로 변환

python-docx를 이용하여 문서 구조와 스타일 정보를 추출한다.

기능:
1. 문서 내용을 IR 블록으로 변환
2. 각 단락의 원본 스타일 정보를 JSON으로 반환 (인스펙터용)

Word 스타일 → IR 매핑:
  Heading 1  → heading1
  Heading 2  → heading2
  Heading 3  → heading3
  Heading 4  → heading4
  Normal     → body
  Table      → table
"""

from typing import List, Dict, Any, Optional, Tuple
from docx import Document
from docx.shared import Pt, Emu
from docx.enum.text import WD_ALIGN_PARAGRAPH
import os


# Word 스타일 이름 → IR 타입 매핑
STYLE_MAP = {
    "Heading 1": "heading1",
    "Heading 2": "heading2",
    "Heading 3": "heading3",
    "Heading 4": "heading4",
    "Title": "heading1",
    "Subtitle": "heading2",
    "Normal": "body",
    "Body Text": "body",
    "List Paragraph": "body",
    # 한글 스타일명
    "제목 1": "heading1",
    "제목 2": "heading2",
    "제목 3": "heading3",
    "제목 4": "heading4",
    "본문": "body",
}


def parse_docx(file_path: str) -> Tuple[List[Dict[str, Any]], List[Dict[str, Any]]]:
    """
    .docx 파일을 파싱하여 (IR 블록 리스트, 스타일 정보 리스트) 튜플을 반환한다.

    Returns:
        (ir_blocks, style_infos)
        - ir_blocks: 변환용 중간표현
        - style_infos: 인스펙터 표시용 원본 스타일 정보
    """
    doc = Document(file_path)
    ir_blocks: List[Dict[str, Any]] = []
    style_infos: List[Dict[str, Any]] = []

    for element in doc.element.body:
        tag = element.tag.split('}')[-1] if '}' in element.tag else element.tag

        if tag == 'p':
            # 일반 단락
            para = _find_paragraph(doc, element)
            if para is None:
                continue

            ir_type = _map_style(para.style.name if para.style else "Normal")
            runs = _extract_runs(para)

            # 빈 단락도 포함 (줄바꿈 용도)
            ir_blocks.append({"type": ir_type, "runs": runs})
            style_infos.append(_extract_style_info(para))

        elif tag == 'tbl':
            table = _find_table(doc, element)
            if table is None:
                continue

            table_block = _parse_table(table)
            ir_blocks.append(table_block)
            style_infos.append({"type": "table", "rows": len(table.rows), "cols": len(table.columns)})

    return ir_blocks, style_infos


def _find_paragraph(doc: Document, element) -> Optional[Any]:
    """element에 대응하는 Paragraph 객체를 찾는다."""
    for para in doc.paragraphs:
        if para._element is element:
            return para
    return None


def _find_table(doc: Document, element) -> Optional[Any]:
    """element에 대응하는 Table 객체를 찾는다."""
    for table in doc.tables:
        if table._element is element:
            return table
    return None


def _map_style(style_name: str) -> str:
    """Word 스타일 이름을 IR 타입으로 변환한다."""
    return STYLE_MAP.get(style_name, "body")


def _extract_runs(para) -> List[Dict[str, Any]]:
    """단락의 run들을 IR run 리스트로 변환한다."""
    runs = []
    for run in para.runs:
        text = run.text
        if not text:
            continue
        runs.append({
            "text": text,
            "bold": bool(run.bold),
            "italic": bool(run.italic) if run.italic else False,
            "underline": bool(run.underline) if run.underline else False,
            "color": _color_to_hex(run.font.color.rgb) if run.font.color and run.font.color.rgb else None,
        })

    if not runs and para.text:
        runs.append({"text": para.text, "bold": False})

    return runs


def _extract_style_info(para) -> Dict[str, Any]:
    """단락의 원본 스타일 정보를 추출한다 (인스펙터용)."""
    style_info: Dict[str, Any] = {
        "style_name": para.style.name if para.style else "Normal",
        "ir_mapping": _map_style(para.style.name if para.style else "Normal"),
        "text_preview": para.text[:50] if para.text else "",
    }

    # 단락 서식
    pf = para.paragraph_format
    if pf:
        align_map = {
            WD_ALIGN_PARAGRAPH.LEFT: "left",
            WD_ALIGN_PARAGRAPH.CENTER: "center",
            WD_ALIGN_PARAGRAPH.RIGHT: "right",
            WD_ALIGN_PARAGRAPH.JUSTIFY: "justify",
        }
        style_info["paragraph"] = {
            "alignment": align_map.get(pf.alignment, "unknown") if pf.alignment is not None else None,
            "line_spacing": float(pf.line_spacing) if pf.line_spacing else None,
            "space_before_pt": float(pf.space_before.pt) if pf.space_before else None,
            "space_after_pt": float(pf.space_after.pt) if pf.space_after else None,
            "first_line_indent_pt": float(pf.first_line_indent.pt) if pf.first_line_indent else None,
            "left_indent_pt": float(pf.left_indent.pt) if pf.left_indent else None,
        }

    # 첫 번째 run의 글자 서식
    if para.runs:
        run = para.runs[0]
        font = run.font
        style_info["font"] = {
            "name": font.name,
            "size_pt": float(font.size.pt) if font.size else None,
            "bold": font.bold,
            "italic": font.italic,
            "underline": font.underline is not None and font.underline is not False,
            "color": _color_to_hex(font.color.rgb) if font.color and font.color.rgb else None,
        }

    return style_info


def _parse_table(table) -> Dict[str, Any]:
    """Word 표를 IR table 블록으로 변환한다."""
    rows = []
    for row in table.rows:
        cells = []
        for cell in row.cells:
            runs = []
            for para in cell.paragraphs:
                if runs:
                    # 단락 사이 줄바꿈
                    runs.append({"text": "\n", "bold": False})
                if para.runs:
                    for r in para.runs:
                        if not r.text:
                            continue
                        runs.append({
                            "text": r.text,
                            "bold": bool(r.bold),
                        })
                elif para.text:
                    runs.append({"text": para.text, "bold": False})
            if not runs:
                runs.append({"text": "", "bold": False})
            cells.append({"runs": runs})
        rows.append(cells)

    # 첫 행이 헤더인지 추정 (첫 행의 모든 run이 bold이면 헤더)
    has_header = False
    if rows:
        first_row = rows[0]
        if all(
            all(r.get("bold", False) for r in cell.get("runs", []) if r.get("text", "").strip())
            for cell in first_row
            if any(r.get("text", "").strip() for r in cell.get("runs", []))
        ):
            has_header = True

    return {"type": "table", "rows": rows, "has_header": has_header}


def _color_to_hex(rgb) -> Optional[str]:
    """RGBColor를 #RRGGBB 문자열로 변환한다."""
    if rgb is None:
        return None
    return f"#{rgb}"


def get_style_report(file_path: str) -> List[Dict[str, Any]]:
    """
    인스펙터용: .docx 파일의 모든 단락에 대한 스타일 정보만 반환한다.
    """
    _, style_infos = parse_docx(file_path)
    return style_infos

File: _legacy/HWFlow_legacy/js-engine/src/parser_docx.js
/**
 * parser_docx.js — .docx 파일을 중간표현(IR)으로 변환
 *
 * fflate로 ZIP 해제, fast-xml-parser로 XML 파싱.
 * python-docx 대체.
 *
 * Word 스타일 → IR 매핑:
 *   Heading 1  → heading1
 *   Heading 2  → heading2
 *   Heading 3  → heading3
 *   Heading 4  → heading4
 *   Normal     → body
 *   Table      → table
 */

import { unzipSync } from 'fflate';
import { XMLParser } from 'fast-xml-parser';
import { base64ToUint8Array } from './utils.js';

// Word 스타일 ID/이름 → IR 타입 매핑
const STYLE_MAP = {
  'Heading1': 'heading1',
  'Heading2': 'heading2',
  'Heading3': 'heading3',
  'Heading4': 'heading4',
  'Heading 1': 'heading1',
  'Heading 2': 'heading2',
  'Heading 3': 'heading3',
  'Heading 4': 'heading4',
  'Title': 'heading1',
  'Subtitle': 'heading2',
  'Normal': 'body',
  'BodyText': 'body',
  'Body Text': 'body',
  'ListParagraph': 'body',
  'List Paragraph': 'body',
  // 한글 스타일명
  '제목 1': 'heading1',
  '제목 2': 'heading2',
  '제목 3': 'heading3',
  '제목 4': 'heading4',
  '본문': 'body',
};

// fast-xml-parser 설정
const xmlParserOptions = {
  ignoreAttributes: false,
  attributeNamePrefix: '@_',
  removeNSPrefix: false,
  preserveOrder: false,
  trimValues: false,
  isArray: (name) => ['w:p', 'w:r', 'w:tr', 'w:tc', 'w:t', 'w:tbl'].includes(name),
};


/**
 * .docx 파일(base64)을 파싱하여 IR 블록 리스트를 반환한다.
 * @param {string} base64Data - base64 인코딩된 docx 파일
 * @returns {Array<Object>} IR 블록 리스트
 */
export function parseDocx(base64Data) {
  const zipData = base64ToUint8Array(base64Data);
  const files = unzipSync(zipData);

  // word/document.xml 추출
  const docXmlBytes = files['word/document.xml'];
  if (!docXmlBytes) return [];

  const decoder = _getDecoder();
  const docXmlStr = decoder(docXmlBytes);

  // word/styles.xml에서 스타일 ID → 이름 매핑 구축
  const styleIdToName = {};
  const stylesXmlBytes = files['word/styles.xml'];
  if (stylesXmlBytes) {
    _buildStyleIdMap(decoder(stylesXmlBytes), styleIdToName);
  }

  const parser = new XMLParser(xmlParserOptions);
  const doc = parser.parse(docXmlStr);

  const body = _dig(doc, 'w:document', 'w:body');
  if (!body) return [];

  const blocks = [];

  // body 내의 요소 순회 (w:p, w:tbl)
  const paragraphs = body['w:p'] || [];
  const tables = body['w:tbl'] || [];

  // 순서 유지를 위해 원본 XML의 순서대로 처리해야 하지만
  // fast-xml-parser는 같은 이름의 형제를 배열로 그룹화한다.
  // preserveOrder: true 를 쓰면 순서 보존되지만 접근이 복잡해진다.
  // 대안: preserveOrder 모드로 재파싱
  const orderedBlocks = _parseBodyOrdered(docXmlStr, styleIdToName);
  return orderedBlocks;
}


/**
 * preserveOrder 모드로 body를 파싱하여 요소 순서를 유지한다.
 */
function _parseBodyOrdered(docXmlStr, styleIdToName) {
  const parser = new XMLParser({
    ...xmlParserOptions,
    preserveOrder: true,
  });
  const ordered = parser.parse(docXmlStr);

  // preserveOrder에서는 [{tagName: [...children], ':@': {attrs}}] 형태
  const docNode = _findOrdered(ordered, 'w:document');
  if (!docNode) return [];
  const bodyNode = _findOrdered(docNode['w:document'], 'w:body');
  if (!bodyNode) return [];

  const blocks = [];
  const bodyChildren = bodyNode['w:body'] || [];

  for (const child of bodyChildren) {
    if (child['w:p'] !== undefined) {
      const block = _parseParagraphOrdered(child, styleIdToName);
      if (block) blocks.push(block);
    } else if (child['w:tbl'] !== undefined) {
      const block = _parseTableOrdered(child, styleIdToName);
      if (block) blocks.push(block);
    }
  }

  return blocks;
}

function _findOrdered(arr, tagName) {
  if (!Array.isArray(arr)) return null;
  for (const item of arr) {
    if (item[tagName] !== undefined) return item;
  }
  return null;
}

function _parseParagraphOrdered(pNode, styleIdToName) {
  const pChildren = pNode['w:p'] || [];
  let irType = 'body';
  const runs = [];

  for (const child of pChildren) {
    if (child['w:pPr'] !== undefined) {
      // 단락 속성에서 스타일 추출
      const pPrChildren = child['w:pPr'] || [];
      const pStyleNode = _findOrdered(pPrChildren, 'w:pStyle');
      if (pStyleNode) {
        const attrs = pStyleNode[':@'] || {};
        const styleId = attrs['@_w:val'] || '';
        irType = _mapStyle(styleId, styleIdToName);
      }
    } else if (child['w:r'] !== undefined) {
      // Run 처리
      const run = _parseRunOrdered(child);
      if (run) runs.push(run);
    }
  }

  return { type: irType, runs: runs.length > 0 ? runs : [{ text: '', bold: false }] };
}

function _parseRunOrdered(rNode) {
  const rChildren = rNode['w:r'] || [];
  let bold = false;
  let italic = false;
  let underline = false;
  let color = null;
  let text = '';

  for (const child of rChildren) {
    if (child['w:rPr'] !== undefined) {
      const rPrChildren = child['w:rPr'] || [];
      for (const prop of rPrChildren) {
        if (prop['w:b'] !== undefined) bold = true;
        if (prop['w:i'] !== undefined) italic = true;
        if (prop['w:u'] !== undefined) underline = true;
        if (prop['w:color'] !== undefined) {
          const attrs = prop[':@'] || {};
          const val = attrs['@_w:val'];
          if (val && val !== 'auto') color = '#' + val;
        }
      }
    } else if (child['w:t'] !== undefined) {
      const tContent = child['w:t'];
      if (Array.isArray(tContent)) {
        for (const t of tContent) {
          if (typeof t === 'string') text += t;
          else if (t['#text'] !== undefined) text += t['#text'];
        }
      } else if (typeof tContent === 'string') {
        text += tContent;
      } else if (tContent && tContent['#text'] !== undefined) {
        text += tContent['#text'];
      }
    }
  }

  if (!text) return null;
  const run = { text, bold };
  if (italic) run.italic = italic;
  if (underline) run.underline = underline;
  if (color) run.color = color;
  return run;
}

function _parseTableOrdered(tblNode, styleIdToName) {
  const tblChildren = tblNode['w:tbl'] || [];
  const rows = [];

  for (const child of tblChildren) {
    if (child['w:tr'] !== undefined) {
      const trChildren = child['w:tr'] || [];
      const cells = [];

      for (const tcNode of trChildren) {
        if (tcNode['w:tc'] !== undefined) {
          const cell = _parseCellOrdered(tcNode, styleIdToName);
          cells.push(cell);
        }
      }

      if (cells.length > 0) rows.push(cells);
    }
  }

  // 첫 행이 헤더인지 추정 (첫 행의 모든 run이 bold)
  let hasHeader = false;
  if (rows.length > 0) {
    const firstRow = rows[0];
    hasHeader = firstRow.every(cell => {
      const contentRuns = (cell.runs || []).filter(r => (r.text || '').trim());
      return contentRuns.length === 0 || contentRuns.every(r => r.bold);
    });
  }

  return { type: 'table', rows, has_header: hasHeader };
}

function _parseCellOrdered(tcNode, styleIdToName) {
  const tcChildren = tcNode['w:tc'] || [];
  const runs = [];

  for (const child of tcChildren) {
    if (child['w:p'] !== undefined) {
      if (runs.length > 0) {
        runs.push({ text: '\n', bold: false });
      }
      const pChildren = child['w:p'] || [];
      for (const pChild of pChildren) {
        if (pChild['w:r'] !== undefined) {
          const run = _parseRunOrdered(pChild);
          if (run) runs.push(run);
        }
      }
    }
  }

  if (runs.length === 0) {
    runs.push({ text: '', bold: false });
  }

  return { runs };
}


/**
 * word/styles.xml에서 스타일 ID → 이름 매핑 구축
 */
function _buildStyleIdMap(stylesXml, map) {
  const parser = new XMLParser({
    ignoreAttributes: false,
    attributeNamePrefix: '@_',
    removeNSPrefix: false,
    isArray: (name) => ['w:style'].includes(name),
  });
  const parsed = parser.parse(stylesXml);
  const styles = _dig(parsed, 'w:styles', 'w:style') || [];
  for (const style of styles) {
    const id = style['@_w:styleId'] || '';
    const nameNode = style['w:name'];
    const name = nameNode ? (nameNode['@_w:val'] || '') : '';
    if (id && name) map[id] = name;
  }
}


function _mapStyle(styleId, styleIdToName) {
  // 먼저 ID로 직접 매핑
  if (STYLE_MAP[styleId]) return STYLE_MAP[styleId];
  // ID → 이름으로 변환 후 매핑
  const name = styleIdToName[styleId] || '';
  if (STYLE_MAP[name]) return STYLE_MAP[name];
  return 'body';
}


function _dig(obj, ...keys) {
  let current = obj;
  for (const key of keys) {
    if (current == null) return null;
    current = current[key];
  }
  return current;
}


function _getDecoder() {
  if (typeof TextDecoder !== 'undefined') {
    const decoder = new TextDecoder('utf-8');
    return (bytes) => decoder.decode(bytes);
  }
  // 수동 UTF-8 디코딩 (JSCore 폴리필)
  return (bytes) => {
    let str = '';
    let i = 0;
    while (i < bytes.length) {
      let c = bytes[i];
      if (c < 0x80) { str += String.fromCharCode(c); i++; }
      else if (c < 0xE0) { str += String.fromCharCode(((c & 0x1F) << 6) | (bytes[i+1] & 0x3F)); i += 2; }
      else if (c < 0xF0) { str += String.fromCharCode(((c & 0x0F) << 12) | ((bytes[i+1] & 0x3F) << 6) | (bytes[i+2] & 0x3F)); i += 3; }
      else {
        const cp = ((c & 0x07) << 18) | ((bytes[i+1] & 0x3F) << 12) | ((bytes[i+2] & 0x3F) << 6) | (bytes[i+3] & 0x3F);
        str += String.fromCodePoint(cp);
        i += 4;
      }
    }
    return str;
  };
}


/**
 * .docx 파일(base64)의 스타일 정보를 반환한다 (인스펙터용).
 * @param {string} base64Data
 * @returns {Array<Object>} 스타일 정보 리스트
 */
export function getDocxStyleReport(base64Data) {
  // 간소화 버전: parseDocx 호출 후 각 블록의 타입 정보만 반환
  const blocks = parseDocx(base64Data);
  return blocks.map(b => ({
    type: b.type,
    ir_mapping: b.type,
    text_preview: b.type === 'table'
      ? `[표 ${(b.rows || []).length}행]`
      : (b.runs || []).map(r => r.text || '').join('').slice(0, 50),
  }));
}
